multidb/pinning.py

```diff
@@ -45,13 +45,29 @@ def decorator(*args, **kw):
     def __enter__(self):
         _locals.old = getattr(_locals, 'old', [])
         _locals.old.append(this_thread_is_pinned())
-        pin_this_thread()
 
     def __exit__(self, type, value, tb):
-        if not _locals.old.pop():
+        previous_state = _locals.old.pop()
+        if previous_state:
```

Comment:
ℹ️ This introduces a subtle behaviour change. Previously, manual use of `unpin_this_thread()` would leak out of the context block, for example when:

1. a thread is pinned
2. enters the context block (`_locals.old = True`)
3. is manually unpinned (with `unpin_this_thread()`)
4. exits the context block (`not _locals.old == False` so no action is taken)
5. the thread will be unpinned when it reaches here

This new code will always restore the previous pinning state when exiting a context block. In my view this is more predictable, but I'm open to your opinion on this.